zrml/parimutuel/src/lib.rs

```diff
@@ -0,0 +1,511 @@
+                actual_payoff,
+                sender: who.clone(),
+            });
+
+            Ok(())
+        }
+
+        /// Refund the collateral of losing categorical outcome assets
+        /// in case that there was no account betting on the winner outcome.
+        ///
+        /// # Arguments
+        ///
+        /// - `refund_asset`: The outcome asset to refund.
+        ///
+        /// Complexity: `O(log(n))``, where `n` is the number of categorical assets the market can have.
+        #[pallet::call_index(2)]
+        #[pallet::weight(T::WeightInfo::refund_pot())]
+        #[frame_support::transactional]
+        pub fn refund_pot(origin: OriginFor<T>, refund_asset: AssetOf<T>) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            let market_id = match refund_asset {
+                Asset::ParimutuelShare(market_id, _) => market_id,
+                _ => return Err(Error::<T>::NotParimutuelOutcome.into()),
+            };
+            let market = T::MarketCommons::market(&market_id)?;
+            ensure!(market.status == MarketStatus::Resolved, Error::<T>::MarketIsNotResolvedYet);
+            ensure!(market.scoring_rule == ScoringRule::Parimutuel, Error::<T>::InvalidScoringRule);
+            ensure!(
+                matches!(market.market_type, MarketType::Categorical(_)),
+                Error::<T>::OnlyCategoricalMarketsAllowed
+            );
+            let market_assets = Self::outcome_assets(market_id, &market)?;
+            ensure!(
+                market_assets.binary_search(&refund_asset).is_ok(),
+                Error::<T>::InvalidOutcomeAsset
+            );
+            let winning_outcome = market.resolved_outcome.ok_or(Error::<T>::NoResolvedOutcome)?;
+            let pot_account = Self::pot_account(market_id);
+            let (refund_asset, refund_balance) = match winning_outcome {
+                OutcomeReport::Categorical(category_index) => {
+                    let winning_asset = Asset::ParimutuelShare(market_id, category_index);
+                    let outcome_total = T::AssetManager::total_issuance(winning_asset);
+                    ensure!(outcome_total == <BalanceOf<T>>::zero(), Error::<T>::RefundNotAllowed);
+
+                    let refund_balance = T::AssetManager::free_balance(refund_asset, &who);
+                    ensure!(!refund_balance.is_zero(), Error::<T>::RefundableBalanceIsZero);
+                    debug_assert!(
+                        refund_asset != winning_asset,
+                        "Since we were checking the total issuance of the winning asset to be \
+                         zero, if the refund balance is non-zero, then the winning asset can't be \
+                         the refund asset!"
+                    );
+
+                    (refund_asset, refund_balance)
+                }
+                OutcomeReport::Scalar(_) => return Err(Error::<T>::ScalarMarketsNotAllowed.into()),
+            };
```

Comment:
This was already fixed. It looks like this now:

```rust
        fn do_refund_pot(who: T::AccountId, refund_asset: AssetOf<T>) -> DispatchResult {
            let market_id = match refund_asset {
                Asset::ParimutuelShare(market_id, _) => market_id,
                _ => return Err(Error::<T>::NotParimutuelOutcome.into()),
            };
            let market = T::MarketCommons::market(&market_id)?;
            ensure!(market.status == MarketStatus::Resolved, Error::<T>::MarketIsNotResolvedYet);
            ensure!(market.scoring_rule == ScoringRule::Parimutuel, Error::<T>::InvalidScoringRule);
            ensure!(
                matches!(market.market_type, MarketType::Categorical(_)),
                Error::<T>::NotCategorical
            );
            Self::market_assets_contains(&market, &refund_asset)?;
            let winning_outcome = market.resolved_outcome.ok_or(Error::<T>::NoResolvedOutcome)?;
            let pot_account = Self::pot_account(market_id);
            let winning_asset = match winning_outcome {
                OutcomeReport::Categorical(category_index) => {
                    Asset::ParimutuelShare(market_id, category_index)
                }
                OutcomeReport::Scalar(_) => return Err(Error::<T>::NotCategorical.into()),
            };
            let outcome_total = T::AssetManager::total_issuance(winning_asset);
            ensure!(outcome_total == <BalanceOf<T>>::zero(), Error::<T>::RefundNotAllowed);

            let refund_balance = T::AssetManager::free_balance(refund_asset, &who);
            ensure!(!refund_balance.is_zero(), Error::<T>::RefundableBalanceIsZero);
            debug_assert!(
                refund_asset != winning_asset,
                "Since we were checking the total issuance of the winning asset to be zero, if \
                 the refund balance is non-zero, then the winning asset can't be the refund asset!"
            );

            let slashable_asset_balance = refund_balance;
            T::AssetManager::slash(refund_asset, &who, slashable_asset_balance);

            let pot_total = T::AssetManager::free_balance(market.base_asset, &pot_account);
            debug_assert!(
                pot_total >= refund_balance,
                "The pot total should be at least as high as the individual refund balance!"
            );
            let refund_balance = refund_balance.min(pot_total);

            T::AssetManager::transfer(market.base_asset, &pot_account, &who, refund_balance)?;

            Self::deposit_event(Event::BalanceRefunded {
                market_id,
                asset: refund_asset,
                refunded_balance: refund_balance,
                sender: who.clone(),
            });

            Ok(())
        }
```